File: utilities/typescript/get_input.ts
import * as fs from 'fs';
import * as path from 'path';

/**
 * Input reading utilities for TypeScript solutions.
 */

/**
 * Read input from input.txt for the given day.
 * @param day Day number (1, 2, 3, etc.)
 * @returns Array of lines as strings
 */
export function getInput(day: number): string[] {
    const filePath = getInputPath(day);
    return readInput(filePath);
}

/**
 * Read input from test_N.txt for the given day and test number.
 * @param day Day number (1, 2, 3, etc.)
 * @param testNum Test number (1, 2, 3, etc.)
 * @returns Array of lines as strings
 */
export function getTestInput(day: number, testNum: number): string[] {
    const filePath = getTestInputPath(day, testNum);
    return readInput(filePath);
}

/**
 * Return the path to input.txt for the given day.
 * Solutions are in days/day_NN/solutions/
 * Input is in days/day_NN/data/input.txt
 */
export function getInputPath(day: number): string {
    return "../data/input.txt";
}

/**
 * Return the path to test_N.txt for the given day and test number.
 */
export function getTestInputPath(day: number, testNum: number): string {
    return `../data/test_${testNum}.txt`;
}

/**
 * Read input file and return lines as an array.
 */
export function readInput(filePath: string): string[] {
    const absPath = path.resolve(filePath);
    return fs.readFileSync(absPath, 'utf-8')
        .trim()
        .split('\n')
        .map(line => line.trim());
}

/**
 * Read input file and return raw content.
 */
export function readInputRaw(filePath: string): string {
    const absPath = path.resolve(filePath);
    return fs.readFileSync(absPath, 'utf-8');
}

File: utilities/elixir/get_input.ex
# Input reading utilities for Elixir solutions

defmodule GetInput do
  # Read input from input.txt for the given day
  def get_input(day) do
    path = get_input_path(day)
    read_input(path)
  end

  # Read input from test_N.txt for the given day and test number
  def get_test_input(day, test_num) do
    path = get_test_input_path(day, test_num)
    read_input(path)
  end

  # Return the path to input.txt for the given day
  def get_input_path(day) do
    "../data/input.txt"
  end

  # Return the path to test_N.txt for the given day and test number
  def get_test_input_path(day, test_num) do
    "../data/test_#{test_num}.txt"
  end

  # Read input file and return lines as a list
  def read_input(file_path) do
    file_path
    |> File.read!()
    |> String.trim()
    |> String.split("\n")
  end

  # Read input file and return raw content
  def read_input_raw(file_path) do
    File.read!(file_path)
  end
end
